### PR TITLE
fix non-nil ptr in struct convert

### LINKED ISSUE #35335: kubectl logs --tail=0 is not working properly
How `docker logs` works:
- `docker logs --tail=-1 CONTAINER_ID`: prints all lines.
- `docker logs --tail=1 CONTAINER_ID`: prints the last line.
- `docker logs --tail=0 CONTAINER_ID`: prints no lines. (Usually used with `-f` to follow new logs)

How `kubectl logs` works:
- `kubectl logs --tail=-1 POD_NAME`: prints all lines.
- `kubectl logs --tail=1 POD_NAME`: prints the last line.
- `kubectl logs --tail=0 POD_NAME`: prints all lines. <-------------- Diffierent.

The reason is that the `tail=0` is ignored when converting the parameters https://github.com/kubernetes/kubernetes/blob/master/pkg/conversion/queryparams/convert.go#L111.

Confirmed with @caesarxuchao.

What is the expected behavior? @smarterclayton 
/cc @kubernetes/sig-api-machinery

### PR DESCRIPTION
**Pull request by wrfly**
**What this PR does / why we need it**:
Fix an issue.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #35335

**Special notes for your reviewer**:
Thanks for other contributer's debug.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go [hunk 1]
```diff
@@ -115,14 +115,17 @@ func customMarshalValue(value reflect.Value) (reflect.Value, bool) {
 }
 
 func addParam(values url.Values, tag string, omitempty bool, value reflect.Value) {
-	if omitempty && zeroValue(value) {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go]: zeroValue should not return true for a *int64 set to 0. what value was getting passed in here?

Comment (wrfly) [reply staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go]: It's `tailLines`, see #35335 
If user set that value to zero, the program should not ignore it, cause he may want to tail 0 lines.

Comment (wrfly) [reply staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go]: Maybe the problem is not here...

https://play.golang.org/p/05D2ZcWVr7i

Just remove the `.Elem()` might works but I have not test it yet: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go#L147

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go]: https://github.com/kubernetes/kubernetes/blob/20934328ce1357073d53e922275204f76488b460/staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go#L176

In that case, where we’ve determined we have a pointer type that is non-nil, when we evolve the actual value to serialize, we should always set omitempty to false

Comment (wrfly) [reply staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go]: Agree, so the changes will be:

```go
if isPointerKind(kind) {
	ft = ft.Elem()
	kind = ft.Kind()
	if !field.IsNil() {
		field = reflect.Indirect(field)
		omitempty = false
	}
}
```

Is that OK?

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go]: That seems right. Would like a test demonstrating the serialization that fails currently so we can make sure this fixes the problem

Comment (wrfly) [reply staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go]: I'll add more unit-tests for that later.

**Comment by liggitt** [comment]
thanks. a couple more questions, then LGTM:
1. did you verify this is handled correctly by the API server and kubelet and actually fixes #35335?
2. did you verify no other parameters started getting included incorrectly as a result of this change? (before and after of query string on `kubectl exec --v=6 ...` and `kubectl log --v=6 ...` commands with minimal options would be helpful)

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60696)

**Comment by wrfly** [comment]
I tested with 
`curl "localhost:8001/api/v1/namespaces/default/pods/nginx-8586cf59-7tfnz/log?tailLines=0&follow"` 
and 
`bin/kubectl logs -f --tail 0 nginx-8586cf59-7tfnz`,
it works fine.

And here the logs of `--v 6`:

```text
root@server:/root/k8s-build/kubernetes# bin/kubectl logs -f --tail 0 nginx-8586cf59-7tfnz --v 6
I0306 09:57:20.764543   34565 loader.go:357] Config loaded from file /root/.kube/config
I0306 09:57:20.770548   34565 restmapper.go:209] Invalidating discovery information
I0306 09:57:20.784336   34565 round_trippers.go:405] GET https://10.8.8.10:6443/api 200 OK in 13 milliseconds
I0306 09:57:20.787769   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis 200 OK in 1 milliseconds
I0306 09:57:20.790609   34565 round_trippers.go:405] GET https://10.8.8.10:6443/api/v1 200 OK in 1 milliseconds
I0306 09:57:20.794389   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/apiregistration.k8s.io/v1beta1 200 OK in 1 milliseconds
I0306 09:57:20.798782   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/extensions/v1beta1 200 OK in 1 milliseconds
I0306 09:57:20.801215   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/apps/v1 200 OK in 0 milliseconds
I0306 09:57:20.804724   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/apps/v1beta2 200 OK in 0 milliseconds
I0306 09:57:20.809003   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/apps/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.812403   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/events.k8s.io/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.816407   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/authentication.k8s.io/v1 200 OK in 0 milliseconds
I0306 09:57:20.820047   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/authentication.k8s.io/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.824073   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/authorization.k8s.io/v1 200 OK in 0 milliseconds
I0306 09:57:20.827446   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/authorization.k8s.io/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.830865   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/autoscaling/v1 200 OK in 0 milliseconds
I0306 09:57:20.835176   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/autoscaling/v2beta1 200 OK in 0 milliseconds
I0306 09:57:20.838806   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/batch/v1 200 OK in 0 milliseconds
I0306 09:57:20.842322   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/batch/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.845730   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/certificates.k8s.io/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.849382   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/networking.k8s.io/v1 200 OK in 0 milliseconds
I0306 09:57:20.853581   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/policy/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.857926   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/rbac.authorization.k8s.io/v1 200 OK in 0 milliseconds
I0306 09:57:20.862331   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/rbac.authorization.k8s.io/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.865983   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/storage.k8s.io/v1 200 OK in 0 milliseconds
I0306 09:57:20.869877   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/storage.k8s.io/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.873072   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/admissionregistration.k8s.io/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.877308   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/apiextensions.k8s.io/v1beta1 200 OK in 0 milliseconds
I0306 09:57:20.880660   34565 round_trippers.go:405] GET https://10.8.8.10:6443/apis/rook.io/v1alpha1 200 OK in 0 milliseconds
I0306 09:57:20.889075   34565 round_trippers.go:405] GET https://10.8.8.10:6443/api/v1/namespaces/default/pods/nginx-8586cf59-7tfnz 200 OK in 1 milliseconds
```

```text
root@server:/home/shouguo/k8s-build/kubernetes# bin/kubectl exec nginx-8586cf59-7tfnz --v 6 pwd
I0306 09:59:08.249545   35656 loader.go:357] Config loaded from file /root/.kube/config
I0306 09:59:08.267498   35656 round_trippers.go:405] GET https://10.8.8.10:6443/api/v1/namespaces/default/pods/nginx-8586cf59-7tfnz 200 OK in 16 milliseconds
I0306 09:59:08.385618   35656 round_trippers.go:405] POST https://10.8.8.10:6443/api/v1/namespaces/default/pods/nginx-8586cf59-7tfnz/exec?command=pwd&container=nginx&container=nginx&stderr=true&stdout=true 101 Switching Protocols in 67 milliseconds
/
root@server:/home/shouguo/k8s-build/kubernetes# bin/kubectl exec nginx-8586cf59-7tfnz --v 6 whoami
I0306 09:59:17.384788   35764 loader.go:357] Config loaded from file /root/.kube/config
I0306 09:59:17.401447   35764 round_trippers.go:405] GET https://10.8.8.10:6443/api/v1/namespaces/default/pods/nginx-8586cf59-7tfnz 200 OK in 15 milliseconds
I0306 09:59:17.514189   35764 round_trippers.go:405] POST https://10.8.8.10:6443/api/v1/namespaces/default/pods/nginx-8586cf59-7tfnz/exec?command=whoami&container=nginx&container=nginx&stderr=true&stdout=true 101 Switching Protocols in 61 milliseconds
root
root@server:/home/shouguo/k8s-build/kubernetes#
```

**Comment by liggitt** [comment]
thanks.
/lgtm

**Pull request merged**